cli/listener/src/main.rs

```diff
@@ -220,35 +218,309 @@ pub fn insert_price(conn: &Connection, exchange_rate: ExchangeRate) -> rusqlite:
+    }
+
+    fn run(mut self) -> ! {
+        loop {
+            self.metrics.polls += 1;
+            match get_exchange_rate(self.config, self.opts) {
+                ListenerResult::ErrSnapshot(err) => {
+                    println!("Error while obtaining on-chain state.");
+                    err.print_pretty();
+                    self.metrics.errors += 1;
+                    self.publish_snapshot(None);
+                    self.sleep_after_error();
+                }
+                ListenerResult::OkListener(state, output) => {
+                    println!("{}", output);
+                    self.publish_snapshot(Some(state));
+                    self.sleep();
+                }
+            }
+        }
+    }
+
+    fn sleep_after_error(&mut self) {
+        // For the sleep time we use exponential backoff with jitter [1]. By taking
+        // the time since the last success as the target sleep time, we get
+        // exponential backoff. We clamp this to ensure we don't wait indefinitely.
+        // 1: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+        let time_since_last_success = self.last_read_success.elapsed();
+        let min_sleep_time = Duration::from_secs_f32(0.2);
+        let max_sleep_time = Duration::from_secs_f32(300.0);
+        let target_sleep_time = time_since_last_success.clamp(min_sleep_time, max_sleep_time);
+        let sleep_time = self
+            .rng
+            .gen_range(Duration::from_secs(0)..target_sleep_time);
+        println!("Sleeping {:?} after error ...", sleep_time);
+        std::thread::sleep(sleep_time);
+    }
+
+    /// Publish a new snapshot that from now on will be served by the http server.
+    ///
+    /// This also updates the block time estimator, if applicable.
+    fn publish_snapshot(&mut self, exchange_rate: Option<ExchangeRate>) {
+        let now = Instant::now();
+
+        if let Some(_exchange_rate) = exchange_rate.as_ref() {
+            self.last_read_success = now;
+        }
+
+        let snapshot = Snapshot {
+            metrics: self.metrics.clone(),
+            exchange_rate,
+        };
+        self.snapshot_mutex
+            .lock()
+            .unwrap()
+            .replace(Arc::new(snapshot));
+    }
+
+    pub fn sleep(&self) {
+        // Sleep until is time to get the next exchange rate.
+        let sleep_time = std::time::Duration::from_secs(self.opts.poll_frequency_seconds as u64);
+        println!(
+            "Sleeping for {:?} after getting the Solido exchange rate",
+            sleep_time
+        );
+        std::thread::sleep(sleep_time);
+    }
+}
+
+#[derive(Clone)]
+struct Metrics {
+    /// Number of times that we checked the price.
+    polls: u64,
+
+    /// Number of times that we tried to get the exchange rate, but encountered an error.
+    errors: u64,
+}
+
+impl Metrics {
+    pub fn write_prometheus<W: std::io::Write>(&self, out: &mut W) -> std::io::Result<()> {
+        Ok(())
```

Comment:
We should write `solido_listener_polls_total` and `solido_listener_errors_total` here then. (Do we call it `solido_listener`? Maybe `solido_pricedb` would be clearer?)